service-base/src/main/java/org/eclipse/hono/service/credentials/CredentialsPlainPasswordHelper.java

```diff
@@ -0,0 +1,100 @@
+/*******************************************************************************
+ * Copyright (c) 2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+
+package org.eclipse.hono.service.credentials;
+
+import io.vertx.core.json.JsonObject;
+import org.eclipse.hono.util.ClearTextPassword;
+import org.eclipse.hono.util.CredentialsConstants;
+
+import java.security.SecureRandom;
+
+/**
+ * Utility methods for hashing of plain text passwords for Credentials endpoint.
+ */
+public class CredentialsPlainPasswordHelper {
+
+    /**
+     * The cryptographically maximal strong salt for SHA-256 in bytes.
+     */
+    private static final int SALT_LENGTH_SHA256 = 32;
+    /**
+     * The cryptographically maximal strong salt for SHA-512 in bytes.
+     */
+    private static final int SALT_LENGTH_SHA512 = 64;
+
+    private static final String DEFAULT_HASH_FUNCTION = CredentialsConstants.HASH_FUNCTION_BCRYPT;
```

Comment:
what about making this configurable in a property?

Reply:
Since we decided to let the client choose the hash function, I suggest to completely remove this and **require** that the client does so. 
One could argue that it is a feature to allow that the client omits the `"hash-function"` field and lets the server decide. But FMPOV this small "feature" does not justify the additional complexity in code, API, documentation, error handling, ... (making this configurable would require to introduce a new `ConfigProperties` class for `CredentialsHttpEndpoint`). 

Another point is, that there is already a constant `DEFAULT_HASH_FUNCTION` in `CredentialsConstants`, which is used for validating passwords and which set to `sha-256`. Defining a second default hash function would IMHO lead to a semantic mess.

Reply:
Most, i.e. > 95%, of the clients will have no clue about hash functions and their security properties. Consequently, a service provider should in-fact not allow clients to provide a hash function at all but instead enforce the usage of a particular (safe) hash function of his own choice. 

I would therefore propose to ignore the *hash-function* property (or even return a 400 if present) and instead *always* use the configured hash function.

Reply:
Not really *change* but *evolve*, I would say ...